First pass at dockerized API

dmohs opened the pull request.

dmohs commented on api/.project/devstart.rb:
```diff
@@ -0,0 +1,49 @@
+  c = Common.new
+  env = c.load_env
+
+  c.run %W{docker network create #{env.namespace}}
```
Adding the database to the same network makes the container name available as a host name.

dmohs commented on api/.project/devstart.rb:
```diff
@@ -0,0 +1,49 @@
+      -v #{ENV["PWD"]}:/w
+      -p 8081:8081
+      #{cname}
+      ./gradlew --no-daemon appengineRun
```
This does not recompile/restart on file changes. Regression?

jmandel replied:
Did this work for you before? (I'm not sure I've tried. https://github.com/GoogleCloudPlatform/gradle-appengine-plugin has some specific comments on hot-reloading.)

dmohs replied:
I should have tried. I had it working on my [Java7 prototype](https://github.com/broadinstitute/aou-api-prototype-java7), but I was doing Gradle compilation and serving in separate processes. As long as I didn't break it, I'm okay with proceeding and fixing it as a follow-up.

dmohs commented on api/src/main/webapp/WEB-INF/appengine-web.xml:
```diff
@@ -1,12 +0,0 @@
-<appengine-web-app xmlns="http://appengine.google.com/ns/1.0">
```
I think this file is autogenerated. It's in our .gitignore.

dmohs commented on api/src/dev/server/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+
+RUN gcloud components install app-engine-java
+
+RUN apk --no-cache add gettext
```
I couldn't find an easily-available image for the `gettext`, package, so I would have had to build our own custom image for that. Since I had to build a custom image anyway, I just built one that contains all of the tools we use for the dev server.

jmandel commented on api/src/dev/server/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+FROM alpine
```
Can you help me understand the logic for this living in `src`?

dmohs replied:
I don't draw a line between code that runs in a system, code used to build/configure a system, code used to help integrate with other systems, etc., since every part is responsible for providing the developer/user experience I want to achieve. I keep root-level items of a repo to a minimum, so I might have a few other things like docs and a central developer script. The goal is that a developer looking at the repo for the first time would naturally have a narrow focus. The `image-folder/Dockerfile` is a convention I stole from somewhere else. Very open to alternatives—I've never had a strong convention for what belongs in `src`.

markfickett commented on api/.project/devstart.rb:
```diff
@@ -0,0 +1,51 @@
+require_relative "common/common"
+
+def startdb()
+  c = Common.new
```
Could you use a more verbose variable name?

dmohs replied:
I guess this stands for, "common things you are likely to want to do in most scripts." I could call it `common` but that doesn't actually seem any more useful than `c`. Is it?

markfickett replied:
I think it is! I think I had decided that `c` probably stood for `container` -- something Docker-specific. A name that lets me know "these are some library functions that David / we wrote" is helpful.

Is there state in a "common"? Looking at https://github.com/dmohs/project-management/blob/master/common.rb I'm reading it as similar to a Java static file with functions -- essentially a namespace. Though `@@commands` is stateful? It's library functions + a stateful argument parser? Is it possible to reference stateless functions outside of an instance?

markfickett commented on api/.project/devstart.rb:
```diff
@@ -0,0 +1,51 @@
+require_relative "common/common"
+
+def startdb()
```
Could you add docstrings (or equivalent)? These are long enough I can't quite scan the function to see quickly what it's doing. Does this just run `mysqld`? Does it also install MySQL if necessary? Does it run db upgrades, or is that in a different step?

dmohs replied:
Crap. I want these commands to be self-documenting. Let me explain a couple of things and see if that changes your mind. First, you probably were expecting this to do more, but I haven't added the database creation scripts yet, so you didn't miss those—they aren't there. Second, if you type `docker run mysql` and you don't have that image, docker automatically pulls it. This is something you get used to quickly. The `%W{}` just says, "split these things by words" (fairly idiomatic ruby) and `#{}` is ruby's string substitution mechanism. So that whole thing is basically just one `docker run` command with a `docker rm` to remove the container when the script exits. So, if you look with fresh eyes, does it still need a docstring?

markfickett replied:
Thanks! I think `startdb` is OK. I'm still not sure what `c` and `cname` are, so I'd rather a longer name there (shouldn't cost us much). And `serve` is both a longer function and a more generic name, so I think it needs either a longer function name or a doc.

I mostly agree that a short method with a clear name doesn't need docs. But another way I use function names is as an index to the file -- I'd like to be able to scan the list of functions [and their docs, assuming the docs are shorter than the function itself] and know _which_ function I should bother reading more of (or, just call). So even if I can read `serve` and understand it pretty quickly, I'd like to be able to read a name like `serveUiAndApi` -- or like `private serve()` which says "don't bother with me unless you really need to."

I think part of what I was missing was overall file structure. It looks like `register_commands` is where I should start (the registered commands are the public interface to this file), and that calls `serve`, and that calls `startDb`. Is there a notion of private/protected/public in Ruby / did I miss a hint that `serve` and `startDb` are only relevant within this file? (I wasn't sure that they weren't being exposed for other modules to call, in which case I'd look for more verbose names.)

(That's a little rambly, but trying to expose my thought process.)

markfickett commented on api/db/setup_local_vars.sh:
```diff
@@ -1,4 +1,4 @@
 source setup_vars.sh
 export WORKBENCH_PASSWORD=${WORKBENCH_PASSWORD:=workbench!pwd}
-export DB_SERVER=${DB_SERVER:=localhost:3306}
```
This file will eventually go away, right -- this is just an intermediate step that keeps everything working?

dmohs replied:
I agree. This is brittle right now because these values have to exactly match those used to start the database and yet they are defined in two separate places. To keep this PR small I might leave that to a follow-up PR, but if so it will be right on this one's heels.

markfickett commented on api/project.rb:
```diff
@@ -0,0 +1,31 @@
+    STDERR.puts "mkdir failed."
+    exit 1
+  end
+  unless system(*%W{git clone https://github.com/dmohs/project-management.git .project/common})
```
What do you think of having this as a git submodule, instead of an ignored auto-download?

dmohs replied:
We've had some nastiness with git submodules in FireCloud and I think it caused me to steer away. A submodule would be more transparent, so I'll try that out.

markfickett commented on api/project.rb:
```diff
@@ -0,0 +1,31 @@
+
+require_relative ".project/common/common"
+Dir.foreach(".project") do |item|
+  unless item =~ /^\.\.?$/ || item == "common"
```
What is this doing?

dmohs replied:
This searches the `.project` folder and loads everything that isn't `.`, `..`, or `common` (the source repo). That lets you dump a new script into that directory and it will be exposed in `./project.rb` without explicitly loading it. Your question makes me think this was a poor tradeoff, especially because these developer scripts change much less frequently than other parts of the repo.

markfickett commented on api/.project/devstart.rb:
```diff
@@ -0,0 +1,51 @@
+require_relative "common/common"
```
I guess `.project` is a convention? I really don't like it; these are important scripts for using/running the project, so putting them in a (conventionally) hidden folder just seems to obscure how it's working. Is there a requirement that we put the scripts in a dot-folder?

jmandel replied:
Piling on: I don't mind dotfiles, but `.project` conflicts with with [a file I already have](https://help.eclipse.org/oxygen/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fmisc%2Fproject_description_file.html) at this path form using some standard dev tools.

blrubenstein replied:
It may be useful to name it something related to what the scrips inside are planned to do. (docker-images or something?)

dmohs replied:
It is hidden because a developer initially approaching the project needs only run `./project.rb` to get a sense of what he/she can do, and only needs to start peeking in to the `.project` folder when he/she needs to make changes to that (likely much later in the ramp-up process than changing code, for example). It also makes tab-completion of `./project.rb` slightly more convenient. I am not in love with this naming convention, so please do suggest alternatives.

dmohs replied:
How about `./project.rb` and `libproject` for the folder? That lets you easily tab-complete `./project.rb` without hiding the helper scripts.

markfickett replied:
`libproject` works for me!

jmandel commented:
Thanks @dmohs! This worked great for me.

I wanted to do the comparison with docker-compose at this stage, so I used the Dockerfile you already created but replaced the ruby with a docker-compose file (recognizing this is less flexible in general). You can see the [diff here](https://github.com/all-of-us/workbench/compare/dmohs/dockerby...jmandel/compare-with-docker-compose?expand=1)

Basic usage is

## Launch the whole stack with logs in the foreground terminal
```
  docker-compose up
```

## Launch the whole stack in the backgroud, and bring API server logs into the foregroud
```
docker-compose up -d
docker-compose up workbench-api
```

## Get a shell that can run/explore the API server 
(This launches dependencies first — namely the db server.)
```
  docker-compose run workbench-api bash
```

dmohs commented:
@jmandel @markfickett Thank you both for taking a pass at this. Take a look at my responses when you can, but I definitely have enough to do another iteration when I get back on Wednesday.

dmohs commented:
@markfickett Thank you for doing another pass and sorry I didn't respond. I'm nearly done addressing feedback but have a few more changes to make. I will let you know when this is ready for another review.

jmandel commented on api/src/dev/server/Dockerfile:
```diff
@@ -0,0 +1,45 @@
+RUN curl https://dl.google.com/cloudsql/cloud_sql_proxy.linux.amd64 \
+  > /usr/local/bin/cloud_sql_proxy && chmod +x /usr/local/bin/cloud_sql_proxy
+
+RUN curl https://raw.githubusercontent.com/Eficode/wait-for/master/wait-for \
```
Might make me feel better to have a commit hash in this url, to know this binary content isn't changing.

dmohs replied:
Yes, good call.

dmohs commented:
I'm happy with where this has landed. PTAL.

dmohs commented:
Merging. @markfickett feel free to make additional comments and I'll address them in a follow-up.
